### DISCUSSION TITLE
tf2 library for ros2_rust

### DISCUSSION THREAD

**Discussion by DS3a** [ros2-rust/ros2_rust#201] [Q&A] [answered]
Is there a tf2 implementation for ros2_rust? or has it been implemented on a separate package? I'm new to this so I'm not sure if the transform library comes with RCL? If it does. Does it have rust bindings?

**Comment by esteve** [comment, marked as answer]
@DS3a thanks for your interest in `ros2-rust`! At the moment, there's no tf2 package for `ros2-rust`, a straightforward version could just wrap the existing C++ implementation in Rust. Would you like to contribute one? We'd be very happy to help you with it.

**Reply by DS3a** [reply]
I'd be glad to contribute a tf2 package for `ros2-rust`... Although I think it might be better to keep most of the code in Rust instead of wrapping it around C++. Especially since the ROS client library does not come with it. Do let me know what you think, and how I should proceed.

**Reply by esteve** [reply]
It'll be tricky to do a full integration of `tf2` and `ros2-rust`, as the infrastructure is not entirely there yet. For example `ros2-rust` does not support actions, clocks or timers yet, which are essential for `tf2`, that's why I suggested wrapping `tf2` for now (here's how it was done for `rcldotnet` https://github.com/IntelligentRoboticsLabs/geometry2/tree/rcldotnet)

If you'd like to implement the full integration, we'd be more than happy to accept contributions for the aforementioned features in `ros2-rust` :slightly_smiling_face:

**Reply by DS3a** [reply]
Ohh, okay. That makes sense. I'll look into adding support for timers (#121). And hopefully move to the other dependencies and move to the full integration after that.

**Reply by Yadunund** [reply]
@DS3a did you get around to implementing the Rust wrapper for tf2?

**Reply by DS3a** [reply]
Nope. Sorry, I got busy with other stuff

**Comment by ljburtz** [comment]
Hello, any update on tf2 for ros2_rust? already deep into implementation with ros2_rust but now realizing we will also need tf2_msgs and can't find it anywhere! any pointer very appreciated (just need the message definition actually)

**Reply by maspe36** [reply]
You should be able to [vendor](https://github.com/ros2-rust/ros2_rust/blob/main/ros2_rust_jazzy.repos#L4) `tf2_msgs` in your workspace just like [we do here in the ros2_rust repo](https://github.com/ros2-rust/ros2_rust/blob/main/examples/minimal_pub_sub/Cargo.toml#L37) for things like `std_msgs`. Then just make sure you've got `tf2_msgs` declared in your Cargo.toml and package.xml so we can do the dependency patching in your `<colcon ws>/.cargo/config.toml`.

Once you vendor `tf2_msgs`, you'll need to build once with `colcon build` and then everything should work even when building with only cargo.

**Reply by ljburtz** [reply]
thank you for the guidance. i could get these tf2_msgs into ros2_rust by: 
adding 
1. the following to `ros2_rust/ros2_rust_humble.repos`

>   ros2/geometry2:
    type: git
    url: https://github.com/ros2/geometry2.git
    version: humble

2. re-importing the repos

> vcs import src < src/ros2_rust/ros2_rust_humble.repos
source /opt/ros/humble/setup.sh
colcon build

3. adding tf2_msgs to the dependencies in my Cargo.toml


Would it make sense to open a PR for adding the update to the ros2_rust_humble.repos?

**Reply by maspe36** [reply]
Awesome! Glad you got it working. No I don't think this needs to go up to ros2_rust. 

Really the crux of this is that in the source of your colcon workspace, you needed a copy of tf2_msgs so that our rust message generation would run. Anywhere in `<your ws>/src` would suffice. That repos file you're mentioning is specifically for ros2 messages that ros2_rust needs for our build. 

Hope this helps 👍

**Reply by MaxiMaerz** [reply]
If this is still needed, I created a new TF implementation which works with this library:

https://github.com/MaxiMaerz/schiebung

The project is still in an early stage and not fully battle tested. However, I'm confident it should cover the basic applications pretty well (e.g. running a local buffer)

Commenting here because it is one of the first hits when you search for tf2 in ros2 with rust.

**Comment by olingo99** [comment]
Hello everyone! I took a crack at what @esteve suggested and this is what I came up with: https://github.com/olingo99/tf2_rs

The basic idea is to expose the tf2 C++ BufferCore via cxx and then build a more idiomatic Rust wrapper on top. It’s still early-stage, but it already supports a local buffer + listener and basic lookup_transform / can_transform usage.

I’m posting it here to make the work public and get feedback from the community before going further: Feel free to recommend changes.

**Reply by romainreignier** [reply]
Nice!
It might be nicer pass a `std::time::Duration` to the `BufferCore` constructor instead of nanoseconds because it needs a lot of zeroes for 10 seconds and it is error prone.
I don't have a usage for it now but it might become useful in the future.

**Comment by deniz-hofmeister** [comment]
Hey everyone, apologies in advance for maybe bumping a dead thread, but I do see activities scattered over multiple years so I'm taking a gamble here. 

I personally have been working on: https://github.com/deniz-hofmeister/transforms
It is a bare transforms library without any transport layers. There are currently 2 wrappers available already that try to make it tf2 style, batteries included. 

Would love to hear your thoughts and inputs. 

Greetings
